src/action/listdir.rs

```diff
@@ -0,0 +1,732 @@
+    use std::time::{Duration, UNIX_EPOCH};
+    use std::os::unix::fs::MetadataExt;
+
+    UNIX_EPOCH.checked_add(Duration::from_secs(metadata.ctime() as u64))
+}
+
+/// Reads a symbolic link, returning the path to the file that the link points to.
+#[cfg(target_os = "linux")]
+fn get_symlink(metadata: &Metadata, file_path: &Path) -> Option<PathBuf> {
+    if !metadata.file_type().is_symlink() { return None; }
+
+    match fs::read_link(file_path) {
+        Ok(file) => Some(file),
+        Err(error) => {
+            warn!("unable to read symlink: {}", error);
+            None
+        }
+    }
+}
+
+/// Fills all fields of `Response` using path to the file.
+#[cfg(target_os = "linux")]
+fn fill_response(file_path: &Path) -> Result<Response, Error> {
+    use std::os::unix::fs::MetadataExt;
+    let metadata = fs::symlink_metadata(file_path)
+        .map_err(Error::ReadPath)?;
+
+    Ok(Response {
+        mode: Some(metadata.mode().into()),
+        ino: Some(metadata.ino() as u32),
+        dev: Some(metadata.dev() as u32),
+        nlink: Some(metadata.nlink() as u32),
+        uid: Some(metadata.uid() as u32),
+        gid: Some(metadata.gid() as u32),
+        size: metadata.size(),
+        atime: get_access_time(&metadata),
+        mtime: get_modification_time(&metadata),
+        ctime: get_status_change_time(&metadata),
+        blocks: Some(metadata.blocks() as u32),
+        blksize: Some(metadata.blksize() as u32),
+        rdev: Some(metadata.rdev() as u32),
+        flags_linux: get_linux_flags(file_path),
+        symlink: get_symlink(&metadata, file_path),
+        path: file_path.clone().to_path_buf(),
+        crtime: get_creation_time(&metadata),
+    })
+}
+
+/// Fills all fields of `Response` using path to the file.
+#[cfg(not(target_os = "linux"))]
+fn fill_response(file_path: &Path) -> Result<Response, Error> {
+    let metadata = fs::symlink_metadata(file_path)
+        .map_err(Error::ReadPath)?;
+
+    Ok(Response {
+        size: metadata.len(),
+        atime: get_access_time(&metadata),
+        mtime: get_modification_time(&metadata),
+        path: file_path.clone().to_path_buf(),
+        crtime: get_creation_time(&metadata),
+        ..Default::default()
+    })
+}
+
+pub fn handle<S: Session>(session: &mut S, request: Request)
+                          -> session::Result<()> {
+    let dir_path = &request.path;
+    let mut paths: Vec<PathBuf> = dir_path.read_dir()
+        .map_err(Error::ReadPath)?.filter_map(|entry| entry.ok())
+        .map(|entry| entry.path()).collect();
+    paths.sort();
+
+    for file_path in &paths {
+        session.reply(fill_response(file_path)?)?;
+    }
+
+    Ok(())
+}
+
+/// Constructs `PathBuf` from `String`. If provided string is empty constructs
+/// `PathBuf` from "/".
+fn get_path(path: &Option<String>) -> PathBuf {
+    match path {
+        Some(path) if !path.is_empty() => PathBuf::from(path),
+        _ => PathBuf::from("/"),
```

Comment:
Prefer to use explicit `None` over `_`. In general, it is better to have explicit patterns for enum types, as they will make the compiler tell you that you should revisit it when you add new variants or change types. Using wildcards will usually make it silently compile, but possibly with incorrect behaviour.

Reply:
The `pattern `&Some(_)` not covered` error occurs in that case (`None` instead of `_`). It happens because of `if` condition: `_` here covers empty `path` and `None` variants.

Reply:
Oh, sorry, you are right, of course. I didn't notice the guard clause.